### ISSUE TITLE
It works well on small dataset, but not well on larger dataset.

### ISSUE THREAD

**Issue by jdxyw** [justheuristic/prefetch_generator#7]
Hi

My usage is like below. Use the new one to replace the old dataloader.

```python
class DataLoaderX(data.DataLoader):
    def __iter__(self):
        return BackgroundGenerator(super().__iter__(), max_prefetch=3)
```

it works very well when the dataset is small. In my case, the dataset contains 5000 images. However, when I try to use it on a much larger dataset that contains 2million images.  It works just like the original dataloader. 

Do you know what's the reason?

Best Regards,

**Comment by justheuristic** [comment]
I am afraid that this is the expected behavior.

Pytorch dataloader is decently optimized for loading large datasets from external memory, there isn't much that can be improved there.

In order to optimize performance for large datasets, the first step is to identify the bottleneck. As a rule, it is one of the two:


__1. Compute bottleneck__ - if you run some heavy augmentations on your images, your CPU may be unable to run these augmentations fast enough to feed the ~~beast~~ gpu. In that case, you can try to optimize the augmentation code using specialized libraries like [albumentations](https://github.com/albumentations-team/albumentations) or custom code on GPU.
- Or you can just get more CPU cores, that would work just as well :)
- If you have non-GPU servers connected to your main server with a fast network (at least 1Gbps), you can let these servers run the augmentations for you and feed you with pre-augmented batches.

__2. IO bottleneck__ - if you're reading your images from a HDD / SSD drive, you may be bottlenecked by the speed at which you read them into RAM. In that case, there's a bunch of things that can be done:

- if you use your own hardware with images located on HDD/sataSSD, just buy a cheap NVMe drive. You can find something [like this](https://www.amazon.com/XPG-SX8200-Gen3x4-3000MB-ASX8200PNP-512GT-C/dp/B07K1HMMJC) for around $100 and it will give you ~5x the read performance of even sata3 SSD. Before you choose this option please check __that your mobo has an m.2 NVME slot!__.

- if you use a server or cloud VM, you can get enough RAM to store the data entirely in RAM. For instance, 2M imagenet-sized JPEG images would take 200-250GB, which is within the realm of possibility even for old servers. You can achieve this with [ramdisk/ramfs](https://www.jamescoyle.net/how-to/943-create-a-ram-disk-in-linux) tools or by manually pre-loading custom JPEGs in python (without decompression). I recommend the former.

- if everything else fails, there are dirty software hacks like buffering. You load as many images as you can fit into 90% of free RAM, then sample training batches from these pre-loaded images only. In background, you constantly load more images and once you've loaded K new images, you add them to the main pool (from which you sample training batches) and throw away K old images from the pool. However, be advised:
  - [ ] this can cause your model to require more training batches till convergence, usually by 0-20%
  - [ ] you must read the new images in shuffled order. For best effect, pre-shuffle the entire dataset in advance
  - [ ] if the buffer is _very_ small (e.g. 1% of all data), some layers such as BatchNorm will need re-calibration (re-computing EMA statistics) after the training is done

In either case, you will probably have to write your own code instead of relying on prefetch_generator.

p.s. some of the hacks that i described will require a lot of engineering effort to implement correctly/efficiently. I do __not__ volunteer to help you implement them, please proceed at your own risk.

**Comment by jdxyw** [comment]
Hi @justheuristic , thank you for the detailed explanation. 

In my case, I use a third-pary cloud GPU service that based on k8s. The k8s use the default option for the shared memory for each pod I use, which is 64MB. We can't modify this configuration. This leads to a result, I can't set the `num_workers` option for the dataloader for PyTorch, otherwise, there would be error, just like this one https://github.com/aws/sagemaker-python-sdk/issues/937. Although I have many CPUs in my pod, but only one CPU to load and process the image.  I considered loading all images to the RAM, but I don't have enough RAM.

Anyway, thanks again for your explanation.

**Comment by justheuristic** [comment]
Alas, i can offer naught but condolences.
